odoo/addons/base/models/res_users.py

```diff
@@ -1456,3 +1523,148 @@ def change_password_button(self):
             line.user_id.write({'password': line.new_passwd})
         # don't keep temporary passwords in the database longer than necessary
         self.write({'new_passwd': False})
+
+# API keys support
+API_KEY_SIZE = 20 # in bytes
+INDEX_SIZE = 8 # in hex digits, so 4 bytes, or 20% of the key
+KEY_CRYPT_CONTEXT = passlib.context.CryptContext(
+    # default is 29000 rounds which is 25~50ms, which is probably unnecessary
+    # given in this case all the keys are completely random data: dictionary
+    # attacks on API keys isn't much of a concern
+    ['pbkdf2_sha512'], pbkdf2_sha512__rounds=6000,
+)
+hash_api_key = getattr(KEY_CRYPT_CONTEXT, 'hash', None) or KEY_CRYPT_CONTEXT.encrypt
+class APIKeysUser(models.Model):
+    _inherit = 'res.users'
+
+    api_keys_only_explicit = fields.Boolean()
+    # so it's possible to add or replace
+    api_keys_only = fields.Boolean(compute='_compute_keys_only')
+    api_key_ids = fields.One2many('res.users.apikeys', 'user_id', string="API Keys")
+
+    def __init__(self, pool, cr):
+        init_res = super().__init__(pool, cr)
+        # duplicate list to avoid modifying the original reference
+        type(self).SELF_WRITEABLE_FIELDS = self.SELF_WRITEABLE_FIELDS + ['api_keys_only_explicit', 'api_key_ids']
+        type(self).SELF_READABLE_FIELDS = self.SELF_READABLE_FIELDS + [
+            'api_keys_only_explicit', 'api_keys_only', 'api_key_ids'
+        ]
+        return init_res
+
+    def write(self, vals):
+        r = super().write(vals)
+        # needs to invalidate the UID cache as this potentially makes
+        # previously valid auth invalid
+        if 'api_keys_only_explicit' in vals:
+            self.clear_caches()
+        return r
+
+    @api.depends('api_keys_only_explicit')
+    def _compute_keys_only(self):
+        for u in self:
+            u.api_keys_only = u.api_keys_only_explicit
+
+    def _check_credentials(self, password, user_agent_env):
+        if user_agent_env['interactive']:
+            return super()._check_credentials(password, user_agent_env)
+
+        if not self.env.user.api_keys_only:
+            try:
+                return super()._check_credentials(password, user_agent_env)
+            except AccessDenied:
+                pass
+
+        if self.env['res.users.apikeys']._check_credentials(scope='rpc', key=password) == self.env.uid:
```

Comment:
@odony used "rpc" for the scope of the key-based authentication, I think it doesn't quite match but I can't find a better scope name.